### ISSUE TITLE
Closing all FD's even those which aren't open

### ISSUE THREAD

**Issue by MarcoStuurman** [varnishcache/varnish-cache#2730]
Possibly related commit: https://github.com/varnishcache/varnish-cache/commit/c60b4f81391c6d162543ff4471d86e423e187e66
Possibly related issue: #669

## Expected Behavior
Faster loading of VCL's

## Current Behavior
Trying to close all FD's even if they do not exist
(Code https://github.com/varnishcache/varnish-cache/blob/master/lib/libvarnish/vsub.c#L58)

## Possible Solution
After master starts up keep track of opened FD's and close them when needed. Maybe we could prevent VSUB_closefrom from running for each VCL load. Or maybe this is a solution:

    void
    VSUB_closefrom(int fd)
    {
    
        assert(fd >= 0);

    #ifdef HAVE_CLOSEFROM
            closefrom(fd);
    #else

    DIR* dirFile = opendir("/proc/self/fdinfo");
    int fdClose;
    if (dirFile)
    {
       struct dirent* fdFile;
       while ((fdFile= readdir(dirFile)) != NULL)
       {
           sscanf(fdFile->d_name, "%d", &fdClose);
           if (fdClose > fd) {
               (void)close(fdClose);
           }
       }
       closedir(dirFile);
    }

    //      int i = sysconf(_SC_OPEN_MAX);
    //      assert(i > 0);
    //      for (; i > fd; i--)
    //              (void)close(i);
    #endif
    }

_(I'm not a C developer)_

## Steps to Reproduce (for bugs)
Strace on varnish and load random VCL

## Context
When we've had a server crash we had to load ~1400 VCL's, our NOFILE limit was set to 900.000, it took 0.9 seconds per VCL to load (without compile time) because it was trying to close all 900.000 FD's

## Your Environment
CentOS 7 (also tested on 6)
Varnish 5.2.1 (but code seems to be the same in 6.0)

**Comment by nigoroll** [comment]
On linux, there is libbsd which contains closefrom() with these alternative
implementations:

- fcntl(fd, F_CLOSEM)
- pstat_getproc to get the highest used fd
- /dev/fd or /proc/self/fd to determine open file descriptors
- naive fallback

On my linux, the /proc implementation is used:

slink@haggis:~/Devel/libbsd (master)$ egrep -i 'closem|getproc|dirfd' config.h
/* Define to 1 if you have the `dirfd' function. */
#define HAVE_DIRFD 1
/* Define to 1 if you have fcntl(F_CLOSEM) */
/* #undef HAVE_FCNTL_CLOSEM */
/* Define to 1 if you have the `pstat_getproc' function. */
/* #undef HAVE_PSTAT_GETPROC */

So some obvious options are:

* link libbsd if available

* pull in libbsd's closefrom()

**Comment by nigoroll** [comment]
On linux, there is libbsd which contains `closefrom()` with these alternative implementations:

- fcntl(fd, F_CLOSEM)
- `pstat_getproc()` to get the highest used fd
- `/dev/fd` or `/proc/self/fd` to determine open file descriptors
- naive fallback

On my linux, the `/proc` implementation is used:

```
slink@haggis:~/Devel/libbsd (master)$ egrep -i 'closem|getproc|dirfd' config.h
/* Define to 1 if you have the `dirfd' function. */
#define HAVE_DIRFD 1
/* Define to 1 if you have fcntl(F_CLOSEM) */
/* #undef HAVE_FCNTL_CLOSEM */
/* Define to 1 if you have the `pstat_getproc' function. */
/* #undef HAVE_PSTAT_GETPROC */
```

So some obvious options are:

* link libbsd if available
* pull in libbsd's `closefrom()`

**Comment by bsdphk** [comment]
Leaving aside linux not having closefrom() anywhere nearer than libbsd, which we'd prefer not to make a requirement.

Importing compat code from libbsd would bring GPL taint, so that's a no-go.

The best angle of attack is probably to call VSUB_closefrom() less often or less intrusively.

For instance when VSUB exec(2)'s, marking fd's CLOEXEC could do the job.

However implementing this is pretty "detail-oriented" work and probably not happening fast, as there is an information-leakage angle to this, which makes it important to get it 100% right.

**Comment by MarcoStuurman** [comment]
@bsdphk the possible solution in my original post, couldn't that be used? The code works.

**Comment by nigoroll** [comment]
FTR dug up by @slimhazard  https://sourceware.org/bugzilla/show_bug.cgi?id=10353#c2

**Comment by nigoroll** [comment]
fixed in 14721f5379ea644cec138dc73958ecbfcee09888 / 6ab26126194fc81017c1847925b7ea4fc2b10969
